numba_cuda/numba/cuda/cudadrv/driver.py

```diff
@@ -2594,9 +2593,18 @@ class Linker(metaclass=ABCMeta):
     """Abstract base class for linkers"""
 
     @classmethod
-    def new(cls, max_registers=0, lineinfo=False, cc=None):
+    def new(cls, max_registers=0, lineinfo=False, cc=None, **kwargs):
         if config.CUDA_ENABLE_MINOR_VERSION_COMPATIBILITY:
-            return MVCLinker(max_registers, lineinfo, cc)
+            # TODO: circular
+            from . import runtime
+            driver_ver, runtime_ver = driver.get_version(), runtime.get_version()
+            if driver_ver >= (12, 0) and runtime_ver > driver_ver:
```

Comment:
Rather than this logic, I think we should instead allow the user to specifically configure Numba to use pynvjitlink. I think we could add pynvjitlink as a dependency and think about automatically using it after the introduction of this PR.

Reply:
Do you mean exposing pynvjitlink as opt in/out through the config? Or do you mean setting things up so that if pynvjitlink is present in the environment, then use it?

Reply:
I meant exposing it as opt in/out through the config.

Reply:
I see. Would this flag have some overlap with `CUDA_ENABLE_MINOR_VERSION_COMPATIBILITY` then, or what would be the interplay between those two? If `CUDA_ENABLE_MINOR_VERSION_COMPATIBILITY=False` then would `USE_PYNVJITLINK` (?) have no effect?

Reply:
To try and minimise complexity / confusion, I'd go with:

- Keep `CUDA_ENABLE_MINOR_VERSION_COMPATIBILITY` for CUDA 11
- Add a new flag for enabling the pynvjitlink linker
- Raise an error if `CUDA_ENABLE_MINOR_VERSION_COMPATIBILITY` is set for CUDA 12
- Raise an error if the pynvjitlink linker is enabled for CUDA 11

The reasoning is that pynvjitlink enablement is not a 1:1 mapping with supporting or not supporting MVC - it's also to support LTO, so having a flag relating to behaviour you want (enabling MVC) controlling a component that has multiple behaviours seems like an impedance mismatch - therefore for now I'd prefer if the user said what was to happen (enable / disable pynvjitlink) rather than stating what behaviour they want.

I started making a spreadsheet containing all the configurations (CUDA 11, 12, a future 13 that's not compatible with pynvjitlink from 12, etc) x (driver </=/> runtime version) x (LTO requested) x (MVC needed) x (hardware supported by pynvjitlink version) and a summary of which are valid but the complexity got a bit out of control for a quick exercise, so I'm really intent on "have the user say what to do" ahead of "try and choose what to do based on the user's intent" for now.

Reply:
Added some logic that should reflect this.